### PR TITLE
Do not update cubes not watched by player

### PR DESCRIPTION
**Pull request by Foghrye4**
Random tick causes vines, chest, fluids and other blocks to check their neighbors. 
-> Which cause invocation of world.getBlockState 
-> Which cause invocation of CubeProviderServer.getCube(...) 
-> Which cause invocation of AsyncWorldIOExecutor.syncCubeLoad(...) if cube is not loaded 
-> Which cause invocation of loadCubeAsyncPart. 
Later cube garbage collector unload loaded cube and process starts over again.

This PR fix that.
Cubes loaded by `world.getBlockState(...)` or `world.setBlockState(...)` will not tick, removed by garbage collector or iterated in provider unless they are watched by cube watcher.

### CODE REVIEW

**Comment by Cyclonit** [comment]
Well you beat me to it! Have you done any benchmarks for comparison?

**Comment by Foghrye4** [comment]
@Cyclonit do visualVM counts as benchmark?

**Comment by Cyclonit** [comment]
Somewhat yes. Numbers regarding world loading speed or actual server tick speed would of course be better.

**Comment by Foghrye4** [comment]
Test conditions:
Client. Load world, wait for 2 minutes with inventory open (so game not paused). Standing still. 
World type - custom cubic. Position height 99. Biome - extreme hills.

Modified (this PR) version tick time - 27 ms.
Regular (this repository newest) version tick time - 27 ms.

https://youtu.be/fMzpJ8geRdo?t=3s

**Comment by Cyclonit** [comment]
What if you don't do a static test? Something like:
Load world, wait 1 minute standing still, fly forward for 1 minute.

**Comment by Foghrye4** [comment]
Well at least Forge dedicated server shows significant tick time decrease. From ~30 ms to ~20 ms.
Unfortunately without client connected.

**Comment by Foghrye4** [comment]
With player connected:
PR: Jump up to 130 ms and in a few minutes decrease to ~60 ms.
Regular: 170 and ~90 ms respectively.

It's a win.

**Comment by Cyclonit** [comment]
That is most definitely a win. I haven't done any testing myself yet. From a purely code perspective I don't see anything wrong with this PR.

**Comment by Cyclonit** [comment]
So, some testing done and my results confirm those of Foghrye4.

File: src/main/java/cubicchunks/server/ChunkGc.java [hunk 1]
```diff
@@ -97,7 +100,7 @@ private void verifyColumnConsistency() {
     }
 
     private void chunkGc() {
-        Iterator<Cube> cubeIt = cubeCache.cubesIterator();
+        Iterator<Cube> cubeIt = cubeCache.watchedCubesIterator();
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/ChunkGc.java]: If you only iterate over watched cubes here, non-watched cubes will never get unloaded, which defeats the purpose of chunk GC.

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 2]
```diff
@@ -80,6 +80,7 @@
     @Nonnull private ICubeIO cubeIO;
 
     // TODO: Use a better hash map!
+    @Nonnull private XYZMap<Cube> watchedCubeMap = new XYZMap<>(0.7f, 8000);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: IIRC PlayerCubeMap already contains this information. Why make another map for it?

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]
Generally I think this is the wrong approach. This is what the cube ticket system was supposed to solve. The real solution for fixing random block ticking would be to actually rewrite it with a mixin by actually using a real cube iterator and get rid of the stupid hacks to make vanilla chunk ticking code work. Currently vanilla uses `PlayerCubeMap.chunkIterator`to get iterator for all the columns that contain ticable cubes, and CC overrides `getStorageArrays` to return array of all `ExtendedBlockStorage` instances that should be ticked, based on whether it has it doesn't have cube tickets. Which is not how it works in vanilla and does cause issues, and should be fixed.

Instead `cubeIterator` should be added to `PlayerCubeMap`, and the cube ticking code should use that.

Generally, it should be less hacks and more actual fixes.

**Comment by Foghrye4** [comment]
>Currently vanilla uses PlayerCubeMap.chunkIteratorto get iterator for all the columns that contain ticable cubes, and CC overrides getStorageArrays to return array of all ExtendedBlockStorage instances that should be ticked

Nope. Not since #269

File: src/main/java/cubicchunks/server/ChunkGc.java (hunk 1 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/server/ChunkGc.java]: They will be unloaded if they are became watched. Imagine player surrounded by cached cubes. Inner circle - watched cubes and outer circle - synchronously loaded cubes. When player move, outer circle cubes became watched and when he leave them behind they are unloaded and removed from both maps. So everything is OK.

File: src/main/java/cubicchunks/server/CubeProviderServer.java (hunk 2 above)
Comment (Foghrye4) [reply src/main/java/cubicchunks/server/CubeProviderServer.java]: So you suggest to get cubes from PlayerCubeMap to do cube ticks?

**Comment by Niko-sk2x** [comment]
So it should be changed to reflect vanilla logic, including forge forced chunks. Generally I want to mimic vanilla logic because it makes it easy to update, and makes debugging easier as I can see what the vanilla behavior is. 

And the config option can be changed to something like `FORCE_VANILLA_RANDOM_BLOCK_TICK`, with the partially-broken behavior it is now.

File: src/main/java/cubicchunks/server/ChunkGc.java (hunk 1 above)
Comment (Cyclonit) [reply src/main/java/cubicchunks/server/ChunkGc.java]: But that only affects those cubes which were in the direction they are moving to. As soon as you spawn there will be a sphere of non-ticking cubes around you. Unless you penetrate that sphere in every direction at least once, you will leave cubes behind.

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/server/ChunkGc.java]: This is complicated a lot by the fact that player-generated cubes have `LIGHT` requirement which generates additional 2 cubes radius. As you move forwards, you leave a trail of forever-loaded cubes on the right and on the left. And any mod that accidentally loads a cube it shouldn't load will never get that cube unloaded.

File: src/main/java/cubicchunks/server/CubeProviderServer.java (hunk 2 above)
Comment (Niko-sk2x) [reply src/main/java/cubicchunks/server/CubeProviderServer.java]: Cube watchers specifically. Or actually have `getCubeIterator` method in there, a cubic chunks equivalent of `getChunkIterator`.

**Comment by Niko-sk2x** [comment]
And actually currently cubes laded by things like setBlockState aren't ticked, but the ones on the edges of loaded parts are, causing some more cubes to be ticked (unless there is some other bug somewhere)

**Comment by Foghrye4** [comment]
So replace `watchedCubesMap` with cube watchers map and this PR is OK?

**Comment by Niko-sk2x** [comment]
See [this](https://github.com/OpenCubicChunks/CubicChunks/pull/282#pullrequestreview-63663914) comment

**Comment by Foghrye4** [comment]
Than yes, got it. I ask you just to be sure that I understand you correctly.

**Comment by Foghrye4** [comment]
Done.

File: src/main/java/cubicchunks/server/CubeWatcher.java [hunk 3]
```diff
@@ -60,13 +60,13 @@
...
-            this.cube.getTickets().add(this);
-        }
-    };
+	private final Consumer<Cube> consumer = (c) -> {
+	        this.cube = c;
+	        this.loading = false;
+	        if (this.cube != null) {
+	            this.cube.getTickets().add(this);
+	        }
+	};
```
Comment (Cyclonit) [inline src/main/java/cubicchunks/server/CubeWatcher.java]: formatting

Comment (Foghrye4) [reply src/main/java/cubicchunks/server/CubeWatcher.java]: Formatted.

File: src/main/java/cubicchunks/server/ChunkGc.java [hunk 4]
```diff
@@ -85,7 +87,8 @@ private void verifyColumnConsistency() {
             IColumn storedCol = columnIt.next();
             Collection<Cube> storedColumnCubes = storedCol.getLoadedCubes();
             for (Cube c : storedColumnCubes) {
-                if (cubeCache.getLoadedCube(c.getCoords()) != c) {
+            	CubePos crd = c.getCoords();
+				if (cubeCache.getCube(crd.getX(), crd.getY(), crd.getZ(), Requirement.GET_CACHED) != c) {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/ChunkGc.java]: formatting.

And what is wrong with using getLoadedCube? It's shorter, and more readable.

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 5]
```diff
@@ -194,18 +195,23 @@ public boolean saveChunks(boolean alwaysTrue) {
...
+			Cube cube = watchersIterator.next().getCube();
+			if (cube == null)
+				continue;
+			cube.tickCubeServer(() -> System.currentTimeMillis() - i > 40, rand);
+			if (cube.isEmpty() || !doRandomBlockTicksHere)
+				continue;
+			int randomTickCounter = randomTickSpeed;
+			while (randomTickCounter-- > 0)
+				cube.randomTick(this.world, rand);
+		}
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Formatting

File: src/main/java/cubicchunks/server/PlayerCubeMap.java [hunk 6]
```diff
@@ -792,4 +793,8 @@ boolean cubePosChanged() {
     public Iterator<CubeWatcher> getRandomWrappedCubeWatcherIterator(int seed) {
         return this.cubeWatchers.randomWrappedIterator(seed);
     }
+    
+    public Iterator<CubeWatcher> getCubeWatcherIterator() {
+        return this.cubeWatchers.iterator();
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/PlayerCubeMap.java]: See vanilla logic for `PlayerChunkMap#getChunkIterator` and replicate it for cubes. Either to it right, or leave as is.

Also doing this part right would remove the need for most (all?) of your `GET_WATCHED` requirement usages.

File: src/main/java/cubicchunks/server/PlayerCubeMap.java [hunk 7]
```diff
@@ -774,12 +778,9 @@ CubePos getManagedCubePos() {
...
             // did the player move far enough to matter?
-            double blockDX = blockToCube(playerEntity.posX) - this.getManagedCubePosX();
-            double blockDY = blockToCube(playerEntity.posY) - this.getManagedCubePosY();
-            double blockDZ = blockToCube(playerEntity.posZ) - this.getManagedCubePosZ();
-
-            double distanceSquared = blockDX * blockDX + blockDY * blockDY + blockDZ * blockDZ;
-            return distanceSquared > 0.9;//0.9 instead of 1 because floating-point numbers may be weird
+			return blockToCube(playerEntity.posX) != this.getManagedCubePosX()
+					|| blockToCube(playerEntity.posY) != this.getManagedCubePosY()
+					|| blockToCube(playerEntity.posZ) != this.getManagedCubePosZ();
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/PlayerCubeMap.java]: Formatting.

Also unrelated change.

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 8]
```diff
@@ -596,7 +611,7 @@ boolean tryUnloadCube(Cube cube) {
 
         // unload the Cube!
         cube.onUnload();
-
+
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: unnecessary changes

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 9]
```diff
@@ -326,7 +341,7 @@ public Cube getCube(int cubeX, int cubeY, int cubeZ, Requirement req) {
      */
     private void onCubeLoaded(@Nullable Cube cube, IColumn column) {
         if (cube != null) {
-            cubeMap.put(cube); // cache the Cube
+			cubeMap.put(cube); // cache the Cube
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Formatting

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 10]
```diff
@@ -296,7 +302,16 @@ public void asyncGetCube(int cubeX, int cubeY, int cubeZ, Requirement req, Consu
...
     @Nullable @Override
     public Cube getCube(int cubeX, int cubeY, int cubeZ, Requirement req) {
-        Cube cube = getLoadedCube(cubeX, cubeY, cubeZ);
+    	if (req == Requirement.GET_WATCHED){
+            PlayerCubeMap playerCubeMap = ((PlayerCubeMap)this.world.getPlayerChunkMap());
+    		CubeWatcher cubeWatcher = playerCubeMap.getCubeWatcher(cubeX, cubeY, cubeZ);
+    		if(cubeWatcher == null)
+        		return null;
+    		else
+    			return cubeWatcher.getCube();
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Formatting

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 11]
```diff
@@ -194,18 +195,23 @@ public boolean saveChunks(boolean alwaysTrue) {
...
-        for (Cube cube : cubeMap) {
-            cube.tickCubeServer(() -> System.currentTimeMillis() - i > 40, rand);
-            if (cube.isEmpty() || !doRandomBlockTicksHere)
-                continue;
-            int randomTickCounter = randomTickSpeed;
-            while (randomTickCounter-- > 0)
-                cube.randomTick(this.world, rand);
-        }
+        PlayerCubeMap playerCubeMap = ((PlayerCubeMap)this.world.getPlayerChunkMap());
+        Iterator<CubeWatcher> watchersIterator = playerCubeMap.getCubeWatcherIterator();
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: See vanilla logic for this in `WorldServer#updateBlocks`:
```java
            for (Iterator<Chunk> iterator = getPersistentChunkIterable(this.playerChunkMap.getChunkIterator()); iterator.hasNext(); this.profiler.endSection())
            {
```

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 12]
```diff
@@ -84,6 +84,7 @@
 
     @Nonnull private ICubeGenerator cubeGen;
     @Nonnull private Profiler profiler;
+
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Unnecessary change

File: src/main/java/cubicchunks/network/PacketCubicWorldData.java [hunk 13]
```diff
@@ -117,7 +117,6 @@ public int getMaxGenerationHeight() {
         @Nullable @Override
         public IMessage handleClientMessage(EntityPlayer player, PacketCubicWorldData message, MessageContext ctx) {
             PacketUtils.ensureMainThread(this, player, message, ctx);
-
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/network/PacketCubicWorldData.java]: This is the only change in the file

File: src/main/java/cubicchunks/lighting/FirstLightProcessor.java [hunk 14]
```diff
@@ -325,7 +326,7 @@ private static boolean needsSkylightUpdate(@Nonnull FastCubeBlockAccess access,
      */
     private static boolean canUpdateCube(@Nonnull Cube cube) {
         BlockPos cubeCenter = getCubeCenter(cube);
-        return cube.getCubicWorld().testForCubes(cubeCenter, UPDATE_RADIUS, Objects::nonNull);
+        return cube.getCubicWorld().testForCubes(cubeCenter, UPDATE_RADIUS, Objects::nonNull, Requirement.GET_WATCHED);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/lighting/FirstLightProcessor.java]: FirstLightProcessor should use LOADED cubes, not just watched ones. This is because some non-watched cubes are loaded specifically so that FirstLightProcessor can use them.

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_HeightLimits.java [hunk 15]
```diff
@@ -181,7 +182,8 @@ private void isAreaLoadedInject(int xStart, int yStart, int zStart, int xEnd, in
                 this.testForCubes(
                         xStart, yStart, zStart,
                         xEnd, yEnd, zEnd,
-                        Objects::nonNull);
+                        Objects::nonNull, 
+                        Requirement.GET_WATCHED);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld_HeightLimits.java]: This will break a whole lot of things for no good reason, and it's not the vanilla logic. Mods and vanilla use it to check if an area is LOADED, not if any player needs it.

File: src/main/java/cubicchunks/client/CubeProviderClient.java [hunk 16]
```diff
@@ -152,6 +153,14 @@ public Cube getCube(int cubeX, int cubeY, int cubeZ) {
         }
         return cube;
     }
+    
+	@Override
+	@Nullable
+	public Cube getCube(int cubeX, int cubeY, int cubeZ, Requirement req) {
+		if (req.compareTo(Requirement.GET_CACHED) <= 0)
+			return this.getLoadedCube(cubeX, cubeY, cubeZ);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/client/CubeProviderClient.java]: formatting

File: src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java [hunk 17]
```diff
@@ -164,7 +165,7 @@ public boolean testForCubes(CubePos start, CubePos end, Predicate<Cube> cubeAllo
         for (int cubeX = minCubeX; cubeX <= maxCubeX; cubeX++) {
             for (int cubeY = minCubeY; cubeY <= maxCubeY; cubeY++) {
                 for (int cubeZ = minCubeZ; cubeZ <= maxCubeZ; cubeZ++) {
-                    Cube cube = this.getCubeCache().getLoadedCube(cubeX, cubeY, cubeZ);
+                    Cube cube = this.getCubeCache().getCube(cubeX, cubeY, cubeZ, req);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/asm/mixin/core/common/MixinWorld.java]: This causes very counterintuitive behavior. Passing in Requirement any other than cached or watched will cause the cube to be loaded. And considering my other comments, I doubt it's needed. And if you want to check if a cube is watched, make separate predicate for it, and pass it in as cubeAllowed.

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 18]
```diff
@@ -356,7 +371,7 @@ private Cube postCubeLoadAttempt(int cubeX, int cubeY, int cubeZ, @Nullable Cube
         // change the async load request result, so the cube here will still be null. Just to make sure, get the cube here
         // otherwise we may end up generating the same cube twice
         if (cube == null) {
-            cube = getLoadedCube(cubeX, cubeY, cubeZ);
+            cube = cubeMap.get(cubeX, cubeY, cubeZ);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Why change it?

File: src/main/java/cubicchunks/server/CubeProviderServer.java [hunk 19]
```diff
@@ -276,7 +282,7 @@ public Cube getLoadedCube(CubePos coords) {
      * @see #getCube(int, int, int, Requirement) for the synchronous equivalent to this method
      */
     public void asyncGetCube(int cubeX, int cubeY, int cubeZ, Requirement req, Consumer<Cube> callback) {
-        Cube cube = getLoadedCube(cubeX, cubeY, cubeZ);
+        Cube cube = cubeMap.get(cubeX, cubeY, cubeZ);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/CubeProviderServer.java]: Why change it?

File: src/main/java/cubicchunks/util/FastCubeBlockAccess.java [hunk 20]
```diff
@@ -86,7 +87,7 @@ private FastCubeBlockAccess(ICubicWorld world, ICubeProvider prov, CubePos start
             for (int relativeCubeZ = 0; relativeCubeZ < dz; relativeCubeZ++) {
                 for (int relativeCubeY = 0; relativeCubeY < dy; relativeCubeY++) {
                     Cube cube =
-                            prov.getLoadedCube(originX + relativeCubeX, originY + relativeCubeY, originZ + relativeCubeZ);
+                            prov.getCube(originX + relativeCubeX, originY + relativeCubeY, originZ + relativeCubeZ, Requirement.GET_CACHED);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/util/FastCubeBlockAccess.java]: why change it?

File: src/main/java/cubicchunks/world/ICubeProvider.java [hunk 21]
```diff
@@ -41,6 +42,9 @@
     @Nullable
     Cube getLoadedCube(CubePos coords);
 
+    @Nullable
+	Cube getCube(int cubeX, int cubeY, int cubeZ, Requirement req);
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/world/ICubeProvider.java]: formatting

File: src/main/java/cubicchunks/world/ICubicWorld.java [hunk 22]
```diff
@@ -82,11 +83,12 @@
      * Returns true iff the given Predicate evaluates to true for all cubes for block positions within blockRadius from
      * centerPos. Only cubes that exist are tested. If some cubes within that range aren't loaded - returns false.
      */
-    default boolean testForCubes(BlockPos centerPos, int blockRadius, Predicate<Cube> test) {
+    default boolean testForCubes(BlockPos centerPos, int blockRadius, Predicate<Cube> test, Requirement req) {
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/world/ICubicWorld.java]: Considering my other comments, I doubt it's needed, and the way Requirement is used is very counterintuitive.

As I explained before, checking if cube is watched can be done in the given predicate.

**Review by Niko-sk2x — CHANGES_REQUESTED** [review]
So I finally took a closer look at the changes and this is the result.

File: src/main/java/cubicchunks/server/ChunkGc.java (hunk 4 above)
Comment (Cyclonit) [reply src/main/java/cubicchunks/server/ChunkGc.java]: Can we stick to using coords instead of unnecessary abbreviations like "crd"? The dropped letters don't increase readability at all.

Comment (Niko-sk2x) [reply src/main/java/cubicchunks/server/ChunkGc.java]: And why not just have a method in ICubeProvider (or whatever the getCube method is in) that takes CubePos? (assuming the version with requirement would be kept)

**Comment by Foghrye4** [comment]
Even without emojis and words I can feel your wrath over that raw (in a past) PR, Barteks2x. I feel like I had played on your nerves. Gladly, you are way more patient than LexManos or I was banned from everything already.
I decide to rewrite it from scratch.

File: src/main/java/cubicchunks/server/PlayerCubeMap.java [hunk 23]
```diff
@@ -798,4 +805,50 @@ boolean cubePosChanged() {
...
+    
+    public Iterator<Cube> getCubeIterator() {
+        WorldServer world = this.getWorldServer();
+        final Iterator<CubeWatcher> iterator = this.cubeWatchers.iterator();
+        ImmutableSetMultimap<ChunkPos, Ticket> persistentChunksFor = ForgeChunkManager.getPersistentChunksFor(world);
+        world.profiler.startSection("forcedChunkLoading");
+        final Iterator<Cube> persistentCubesIterator = persistentChunksFor.keys().stream()
+                .filter(Objects::nonNull)
+                .map(input -> ((IColumn)world.getChunkFromChunkCoords(input.x, input.z)).getLoadedCubes())
+                .collect(ArrayList<Cube>::new, (list,cubeCollection)->((ArrayList<Cube>)list).addAll(cubeCollection), (list,cubeList)->((ArrayList<Cube>)list).addAll(cubeList))
```
Comment (Niko-sk2x) [inline src/main/java/cubicchunks/server/PlayerCubeMap.java:817]: is the ArrayList cast necessary?

Comment (Foghrye4) [reply src/main/java/cubicchunks/server/PlayerCubeMap.java:817]: Yes. Due to code limitation of `collect(...)` `list` is a plain object and contain no methods. I could cast to `List` though.

**Review by Niko-sk2x — APPROVED** [review]
